.tree-sitter-lint/tree-sitter-lint-local/src/lib.rs

```diff
@@ -1 +1,126 @@
+    clap::Parser,
+    lsp::{self, LocalLinter},
+    tree_sitter::Tree,
+    tree_sitter_grep::{RopeOrSlice, SupportedLanguage},
+    Args, Config, FileRunContext, FromFileRunContext, FromFileRunContextInstanceProvider,
+    FromFileRunContextInstanceProviderFactory, FromFileRunContextProvidedTypes,
+    FromFileRunContextProvidedTypesOnceLockStorage, MutRopeOrSlice, Plugin, Rule,
+    ViolationWithContext,
+};
+
+pub fn run_and_output() {
+    tree_sitter_lint::run_and_output(
+        args_to_config(Args::parse()),
+        FromFileRunContextInstanceProviderFactoryLocal,
+    );
+}
+
+pub fn run_for_slice<'a>(
+    file_contents: impl Into<RopeOrSlice<'a>>,
+    tree: Option<&Tree>,
+    path: impl AsRef<Path>,
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+pub fn run_fixing_for_slice<'a>(
+    file_contents: impl Into<MutRopeOrSlice<'a>>,
+    tree: Option<&Tree>,
+    path: impl AsRef<Path>,
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_fixing_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+struct LocalLinterConcrete;
+
+impl LocalLinter for LocalLinterConcrete {
+    fn run_for_slice<'a>(
+        &self,
+        file_contents: impl Into<RopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_for_slice(file_contents, tree, path, args, language)
+    }
+
+    fn run_fixing_for_slice<'a>(
+        &self,
+        file_contents: impl Into<MutRopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_fixing_for_slice(file_contents, tree, path, args, language)
+    }
+}
+
+pub async fn run_lsp() {
+    lsp::run(LocalLinterConcrete).await;
+}
+
+fn args_to_config<T: FromFileRunContextInstanceProviderFactory>(args: Args) -> Config<T> {
+    args.load_config_file_and_into_config(all_plugins(), all_standalone_rules())
+}
+
+fn all_plugins<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Plugin<T>> {
+    vec![tree_sitter_lint_plugin_replace_foo_with::instantiate()]
+}
+
+fn all_standalone_rules<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Arc<dyn Rule<T>>> {
+    local_rules::get_rules()
+}
+
+struct FromFileRunContextInstanceProviderFactoryLocal;
+
+impl FromFileRunContextInstanceProviderFactory for FromFileRunContextInstanceProviderFactoryLocal {
+    type Provider<'a> = FromFileRunContextInstanceProviderLocal<'a>;
```

Comment:
By making all of the entry points generic on the factory I avoided some lifetime issues because the factory has no lifetime, it's only once you call its "factory method" that you get an instance-provider that at that point has a fixed/known lifetime (which is the lifetime of the stuff it will store basically, eg tree nodes or source text references)